tools/wake-format/wake-format.cpp

```diff
@@ -0,0 +1,196 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Open Group Base Specifications Issue 7
+#define _XOPEN_SOURCE 700
+#define _POSIX_C_SOURCE 200809L
+
+#include <stdio.h>
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <vector>
+
+#include "gopt/gopt-arg.h"
+#include "gopt/gopt.h"
+#include "parser/cst.h"
+#include "parser/parser.h"
+#include "parser/syntax.h"
+#include "util/diagnostic.h"
+#include "util/file.h"
+
+#define VERSION "0.0.1"
+
+class TerminalReporter : public DiagnosticReporter {
+ public:
+  TerminalReporter() : errors(false), warnings(false) {}
+  bool errors;
+  bool warnings;
+
+ private:
+  std::string last;
+
+  void report(Diagnostic diagnostic) {
+    if (diagnostic.getSeverity() == S_ERROR) errors = true;
+    if (diagnostic.getSeverity() == S_WARNING) warnings = true;
+
+    if (last != diagnostic.getMessage()) {
+      last = diagnostic.getMessage();
+      std::cerr << diagnostic.getLocation() << ": ";
+      if (diagnostic.getSeverity() == S_WARNING) std::cerr << "(warning) ";
+      std::cerr << diagnostic.getMessage() << std::endl;
+    }
+  }
+};
+
+void print_help(const char *argv0) {
+  // clang-format off
+  std::cout << std::endl
+    << "Usage: " << argv0 << " [OPTIONS] [<file> ...]" << std::endl
+    << "  --dry-run  -n     Check if formatting needed, but don't apply it"      << std::endl
+    << "  --help     -h     Print this help message and exit"                    << std::endl
+    << "  --in-place -i     Edit files in place. Default emits file to stdout"   << std::endl
+    << "  --version  -v     Print the version and exit"                          << std::endl
+    << "  --debug    -d     Print debug info while formatting"                   << std::endl
+    << std::endl;
+  // clang-format on
+}
+
+void print_version() { std::cout << "wake-format " << VERSION << std::endl; }
+
+void print_cst(CSTElement node, int depth) {
+  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
+    std::cout << depth << ": ";
+    for (int i = 0; i < depth; i++) {
+      std::cout << "  ";
+    }
+    std::cout << symbolExample(child.id()) << " " << (child.isNode() ? "[Node]" : "[Leaf]");
+    if (child.isNode()) {
+      std::cout << std::endl;
+      print_cst(child, depth + 1);
+      continue;
+    }
+    switch (child.id()) {
+      case TOKEN_ID:
+      case TOKEN_INTEGER:
+      case TOKEN_DOUBLE:
+      case TOKEN_STR_SINGLE:
+      case TOKEN_REG_SINGLE:
+        std::cout << " -> " << child.fragment().segment().str();
+      default:
+        std::cout << std::endl;
+        break;
+    }
+  }
+}
+
+void walk_cst(std::ostream *out, CSTElement node) {
+  for (CSTElement child = node.firstChildElement(); !child.empty(); child.nextSiblingElement()) {
+    if (child.isNode()) {
+      walk_cst(out, child);
+      continue;
+    }
+
+    if (child.id() == TOKEN_KW_MACRO_HERE) {
```

Comment:
Added the `here` -> `@here` refactor because it was pretty trivial but I'm happy to revert if we want to delay the change.

Reply:
nope love it!